### ISSUE TITLE
Extend the metrics API with  functionality to get subverse specific metrics

### ISSUE THREAD

**Issue by PaulBoon** [IQSS/dataverse#6766]
Our clients want to get metrics only for their 'institutional' subverse, like total number of datasets, files and storage size within their subverse.  With the current API this is not possible because the metrics is only for the whole dataverse instance including all it's verses.

**Comment by PaulBoon** [comment]
At DANS we have implemented this in our fork and are going to make this into a PR; code is at: 
https://github.com/DANS-KNAW/dataverse/blob/dans-develop/src/main/java/edu/harvard/iq/dataverse/api/MetricsDans.java
and 
https://github.com/DANS-KNAW/dataverse/blob/dans-develop/src/main/java/edu/harvard/iq/dataverse/metrics/MetricsDansServiceBean.java

**Comment by djbrooke** [comment]
@PaulBoon thanks. We'll be happy to take a look at a PR but I believe that the Dataverse size API under /dataverses can be passed a specific Dataverse (http://guides.dataverse.org/en/latest/api/native-api.html#report-the-data-file-size-of-a-dataverse). I'll defer to @scolapasta or another developer about the desired implementation here, but I don't think we'd want one API under /metrics and another under /dataverses.

Also, are there any UI changes here?

**Comment by PaulBoon** [comment]
The extension API has a lot of endpoints with return json with much more than a simple count. 
I know this code needs some reverse engineering to produce decent documentation, but the basic url contains the alias of dataverse you want metrics on
https://$SERVER/api/info/metrics-dans/{topLevelDvAlias}/
and then it is followed by: 

- dataverses
- dataverses/addedOverTime
- dataverses/byCategory
- datasets
- datasets/bySubject
- datasets/addedOverTime
- files
- tree
- dataverses-report
- datasets-report
- downloads

It is based on the notion that we have a  hierarchy of sub-verses.

**Comment by djbrooke** [comment]
Thanks @PaulBoon, I understand you're adding a lot of additional functionality. Unless there are objections from the technical team, we'll just plan to wait on the PR.

**Comment by donsizemore** [comment]
@PaulBoon thank you for this. our most common requests are for file download count by year and number of datasets published per year. Is the latter included in datasets-report? If not, I can open another issue after your PR.

**Comment by qqmyers** [comment]
Re: 1 above - and the MakeDataCount metrics are different again, providing a count for a given month specified in the call, but nothing cumulative.

**Comment by qqmyers** [comment]
A progress report - for QDR, I've started adapting the existing /info/metrics endpoints to also support per-Dataverse reporting (e.g. via a ?parentAlias=<x> query param), adding MakeDataCount endpoints that work the same way (aggregating up to the supplied date and being per-Dataverse). I'm also starting to add some endpoints to get files by mimetype (count and total size), and unique downloaders per dataset (a metric QDR wants that counts the number of users who've downloaded any files from a dataset that might be a better metric than # of downloads which emphasizes datasets with many files). 

In doing that, I'm pulling in some of the servicebean methods from the DANS code (e.g. for recursively finding child datasets/dataverses and scoping the other queries based on those results.) If there's other work going on, it would be good to coordinate.

I'm also trying to look for duplication. Right now, the datasets and dataverse APIs don't fully duplicate what's in the metrics code (existing, my extensions, and DANS' extensions) but there is some overlap and alternate ways of doing things like getting children recursively (using objects or a direct SQL query) that could perhaps be reduced. 

Planning a PR(s) soon and hoping to get some discussion in the community meeting and/or on the mailer but it's taking longer to get there than I'd hoped so I wanted to report in case there's other activity ongoing.

**Comment by LauraHuisintveld** [comment]
>     2\. I had a hard time understanding the query for datasetsBySubject so I cut/pasted the query there to run in manually via psql and I don't get the same distribution of subjects/dataset count as I see in the search facets (when logged out so the facets show only published datasets). Should they be the same? Are they the same for DANS? My results could just be a mistake on my part, but I wanted to make sure I understand what I should be seeing before I dig further.

Hi @qqmyers , My colleague Paul who opened this issue will be back from holiday next week. In the mean time, I can react to this question. The result of the metrics for 'Dataset by subject' are not always the same numbers as shown in the facets in the User Interface. This may be caused by the following:

- A dataset can have more than one subject applied to.

- Different versions of the same dataset can have different subjects

- Subject(s) chosen for draft datasets are also included, and can differ from the subject of the last published version of the dataset.

**Comment by qqmyers** [comment]
@LauraHuisintveld - belated thanks for the info above - my guess is that the "Different versions of the same dataset can have different subjects" was the main issue I saw. In #7177, I modified the query so it is only reporting for the latest version (consistent with the existing APIs).

@PaulBoon and others - with #7178 going through review - are there changes you'd need to that to cover your use cases? The #7177 issue has a link to a spreadsheet comparing the existing and new API calls that might be the fastest way to see what the PR adds - it's ~ a merge between the existing API and the DANS code with some modification/extension to include the MDC metrics and handle the queries needed for the corresponding metrics page (see https://github.com/IQSS/dataverse-metrics/pull/52).  Are there additional things from the DANS code listed above that should be included (in this PR or a future one)? If a conf. call to discuss would help, let me know.

**Comment by PaulBoon** [comment]
@qqmyers This is great, after upgrading to 5.x we can probably start using the 'installation-level metrics app' at DANS to have our clients view their institutional (sub-verse) specific metrics. 
Our 'home brew' metrics app looks like this: 
![Screenshot 2020-10-29 at 14 25 15](https://user-images.githubusercontent.com/2151103/97580221-4fb5e480-19f3-11eb-8e42-4cc16bdb0dc5.png)

Maybe you can show us a screen dump of what the newmetrics looks like for the installation level view?

**Comment by qqmyers** [comment]
@PaulBoon - there's a pic of the interface at https://github.com/IQSS/dataverse-metrics/pull/52#issuecomment-683821692. I think the first thing I see that's a difference is that the #7178 PR doesn't show draft information. Some of the underlying queries would support that already, and the rest probably could, but I thought limiting things to public would be a useful first step and not make the one PR any bigger.

**Comment by pdurbin** [comment]
> This is great, after upgrading to 5.x we can probably start using the 'installation-level metrics app' at DANS to have our clients view their institutional (sub-verse) specific metrics.

@PaulBoon it looks like you're running a quite recent version now (5.10.1). Have you had a chance to try the metrics app? Should this issue remain open? Thanks.

(Nice homegrown app, by the way.)

**Comment by PaulBoon** [comment]
Hi @pdurbin, the homegrown app (and API extension) was done by @ekoi. 
This is some time ago, but I do remember that I had problems with the metrics app, maybe I should try it again and see what it does for me. Right now I am occupied with other things, maybe next year I will revisits this. 
The main objective for the API extension was to supply information to our institutional sub-verse users, like total sizes and counts.

**Comment by cmbz** [comment]
To focus on the most important features and bugs,     we are closing issues created before 2020 (version 5.0)     that are not new feature requests with the label 'Type: Feature'.    

If you created this issue and you feel the team     should revisit this decision, please reopen the issue and leave a comment.
